homeassistant/components/notify/matrix.py

```diff
@@ -0,0 +1,189 @@
+    vol.Required(CONF_ROOM): cv.string,
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_service(hass, config):
+    """Get the Matrix notification service."""
+    session_file_path = hass.config.path(SESSION_FILE)
+
+    return MatrixNotificationService(
+        config.get(CONF_HOMESERVER),
+        config.get(CONF_ROOM),
+        session_file_path,
+        config.get(CONF_VERIFY_SSL),
+        config.get(CONF_USERNAME, None),
+        config.get(CONF_PASSWORD, None)
+    )
+
+
+# pylint: disable=too-few-public-methods
+class MatrixNotificationService(BaseNotificationService):
+    """Wrapper for the MatrixNotificationClient."""
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, homeserver, default_room, session_file, verify_ssl,
+                 username=None, password=None):
+        """Buffer configuration data for send_message."""
+        self.homeserver = homeserver
+        self.default_room = default_room
+        self.session_file = session_file
+        self.verify_tls = verify_ssl
+        self.username = username
+        self.password = password
+
+    def send_message(self, message, **kwargs):
+        """Wrapper function to instantiate a MatrixNotificationClient."""
+        if kwargs.get(ATTR_TARGET) is None:
+            target_rooms = [self.default_room]
+        else:
+            target_rooms = kwargs.get(ATTR_TARGET)
+
+        send_message(
+            message,
+            self.homeserver,
+            target_rooms,
+            self.session_file,
+            self.verify_tls,
+            self.username,
+            self.password
+        )
+
+
+# pylint: disable=too-many-arguments
+def send_message(message, homeserver, target_rooms, session_file,
+                 verify_tls=True, username=None, password=None):
+    """Do everything thats necessary to send a message to a Matrix room."""
+    from matrix_client.client import MatrixClient, MatrixRequestError
+
+    class MatrixNotificationClient:
+        """Handle Homeserver connection, authentication and messaging."""
+
+        def __init__(self):
+            """Set up a connection to the homeserver."""
+            self.client = None
+            self.token = None
+
+            self.session_file = session_file
+            self.load_token()
+            if self.token:
+                if not self.login_by_token():
+                    if not self.login_by_password():
+                        _LOGGER.error(
+                            'login failed, both token and username/password '
+                            'invalid'
+                        )
+                        return
+            else:
+                if not self.login_by_password():
+                    _LOGGER.error('login failed, username/password invalid')
+                    return
+
+            # work around library issue
+            # https://github.com/matrix-org/matrix-python-sdk/issues/48
+            self.client.api.txn_id = int(time.time() * 1000)
```

Comment:
I've relayed the question upstream and am waiting for a reponse. However I can create another pull request as soon as the new release happens.